### ISSUE TITLE
Reused data crashes

### ISSUE THREAD

**Issue by virtuald** [pthom/cvnp#13]
I'm going to work on a smaller test/reproducer for this, but leaving this here for now in case you have thoughts on how to address this. Basically, if there's a function that takes a mat argument, and returns the same mat, then the returned numpy array makes a capsule to keep the mat alive (which is great), but there's nothing to keep the original numpy array alive as far as I can tell.

This can be seen in `cv::Mat nparray_to_mat(pybind11::array& a)` where the mat is created with the a.mutable_data, but it's not trying to keep it alive. I believe we need some way to notice that the mat is holding a numpy array and return that original object.

I think the opencv python support works around this by having a custom allocator + umat.

* Allocator uses existing numpy memory: https://github.com/opencv/opencv/blob/4.x/modules/python/src2/cv2_numpy.cpp#L13
* Allocator makes new numpy memory: https://github.com/opencv/opencv/blob/4.x/modules/python/src2/cv2_numpy.cpp#L26
* Mat sets the custom allocator: https://github.com/opencv/opencv/blob/4.x/modules/python/src2/cv2_convert.cpp#L301

Reference: https://github.com/robotpy/mostrobotpy/issues/76

**Comment by pthom** [comment]
Hello,

It would help to work on a minimal reproducible example.

As a starter, I'll be using this code as a demonstration:

Here, I define a function that modifies a matrix inplace, and also returns it:

```cpp
cv::Mat multiply_inplace(cv::Mat& m, double factor)
{
    m = m * factor;
    return m;
}

...
    m.def("multiply_inplace", multiply_inplace);
```

When using it from python, we get:

```python
python
Python 3.12.0 (v3.12.0:0fb18b02c8, Oct  2 2023, 09:45:56) [Clang 13.0.0 (clang-1300.0.29.30)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import numpy as np; import cvnp
>>> m = np.eye(3, 3, dtype=np.float32)
>>> n = cvnp.multiply_inplace(m, 2.0)
>>> id(m)
4304590512
>>> id(n)
4315797840
>>> m == n
array([[ True,  True,  True],
       [ True,  True,  True],
       [ True,  True,  True]])
>>> m
array([[2., 0., 0.],
       [0., 2., 0.],
       [0., 0., 2.]], dtype=float32)
>>> 
```

So, yes id(m) is different from id(n). But that seems legit: how could the bindings know that the returned matrix is the same? 

Is this a good repro, or can you provide a closer one?

**Comment by virtuald** [comment]
Here's a reproducer that works in my pybind11 project (which has copies of your stuff from.. may, but I think it's up to date for mat purposes):

```
  // defined in  cscore._cscore 
  m.def("test", [](cv::Mat mat) {
    return mat;
  });
```

t.py:

```
from cscore._cscore import test
import numpy as np
import gc

img = np.zeros(shape=(480, 640, 3), dtype=np.uint8)
for i in range(100):
    img = test(img)
    gc.collect()
    print(img)
```

```
$ python3 t.py 
Segmentation fault (core dumped)
```

**Comment by pthom** [comment]
Your code does not fail on my side, using macOS + python 3.12 + opencv 4.5.5.
I guess more investigation is required, on different OSes + maybe different versions of OpenCV.

**Comment by virtuald** [comment]
Linux + Python 3.8 + Numpy 1.24.4 + OpenCV 4.8 + pybind11 smart_holder branch (2.11+)

... I see my reproducer doesn't crash in your CI either. Weird. Might be something environmental on my side.

Even if this isn't always reproducable, am I missing something about how this works? Where does the original numpy array reference get kept alive by cvnp in `nparray_to_mat`? It seems to me that something should have a reference to it, and I don't see anything.

**Comment by pthom** [comment]
> Where does the original numpy array reference get kept alive by cvnp in nparray_to_mat? It seems to me that something should have a reference to it, and I don't see anything.

I understand your point. I would need to confirm the issue with a repro that triggers an issue however. May be be compiling with asan or something like this?

Adding a custom allocator for cv::Mat, such as done in python/src2/cv2_convert.cpp seems quite challenging and I'm afraid it could lead to other issues if not implemented with extreme care.

At the moment, I can think of possible workaround, such as adding this:

```cpp
    struct KeepAlivePyArrays
    {
        struct MatArrayPair
        {
            cv::Mat* m;
            pybind11::array a;
            double timestamp;
        };
        std::vector<std::pair<cv::Mat*, pybind11::array>> mMatArrayPairs;

        void on_array_to_mat(cv::Mat* m, pybind11::array a)
        {
            // should be called by nparray_to_mat
            mMatArrayPairs.push_back({m, a, now()}
        }
        void on_mat_to_array(cv::Mat* m, pybind11::array a)
        {
            // should be called by mat_to_np_array

            // 1. Suppress item from mMatArrayPairs is m present in mMatArrayPairs
            // (this would make 'a' releasable by decrementing its ref counter)

            // Suppress all instances in mMatArrayPairs which are older than, say, 1 second?
            // (handles cases where nparray_to_mat was called, but never on_array_to_mat for the same mat)
            // this is a sad workaround
        }
    };

```

**Comment by virtuald** [comment]
Both of those workarounds seem insufficient, the custom allocator seems like the Right Way to do this. Should be pretty straightforward, I'll put something together at some point.

**Comment by pthom** [comment]
I took some time to explore the custom allocator, and in fact OpenCV's custom allocator will probably not help.

In the code below:

```cpp
    cv::Mat nparray_to_mat(pybind11::array& a)
    {
        ...
        cv::Mat m(size, type, is_not_empty ? a.mutable_data(0) : nullptr);
        return m;
    }
```

The constructor for cv::Mat does not call the custom allocator, since it does not have to allocate memory for the matrix data (and the content of cv::Mat's structure is allocated on the stack)

We are left with this issue:

- we can call `a.inc_ref()` when constructing the cv::Mat inside nparray_to_mat
- how to make sure that `a.dec_ref()` is called when this particular instance will be destroyed? 

---
Note: in the other direction (`mat_to_nparray`), we do not have the same issue, since we use `detail::make_capsule_mat(m)` which will cause the `cv::Mat` to be destroyed when the `py::array` reference count drops to zero.

**Comment by pthom** [comment]
[This commit](https://github.com/pthom/cvnp/commit/88d903329ded9506d2320fd723126dab6518507b) adds a test that shows the failure. Was fairly complex to setup

**Comment by pthom** [comment]
A possible solution is implemented inside the branch [hunt_seg2](https://github.com/pthom/cvnp/tree/hunt_seg2).

See [related StackOverflow question](https://stackoverflow.com/questions/77990385/complex-c-lifetime-issue-in-python-bindings-between-c-and-numpy/

**Comment by virtuald** [comment]
I haven't tried it yet, but looking at it I don't understand why you added register/unregister instances?

**Comment by virtuald** [comment]
Also, are you sure that `MatAllocator_LinkArray` is ever deleted? The other opencv allocator is statically allocated, so I can't imagine it deletes the allocator pointer.

**Comment by pthom** [comment]
> Also, are you sure that MatAllocator_LinkArray is ever deleted?

There are two tricky points:

- We can't have a global allocator, since it needs to be aware of the py::array, so we need to allocate one per py::array
- In order to make sure it is deleted, we hook the destructor of cv::Mat so that the destruction of cv::Mat also destroys it allocator

See comments in the code:
```cpp
     //
        // This class is a MatAllocator that uses a numpy array as the data pointer
        // --------------------------------------------------------------------------------------------
        // The implementation is quite tricky:
        // - A collection of all instances is kept in a static vector (see register_instance and unregister_instance)
        // - The constructor:
        //     * keeps a reference to the numpy array
        //     * registers itself via register_instance()
        // - allocate:
        //     * creates a new UMatData object
        //     "steals" the data pointer from the numpy array
        // - deallocate will:
        //     * decrements the reference count and deletes the UMatData object if no more references
        //     * unregisters itself via unregister_instance() which will *destroy* this instance!
        //  *The destructor is called via deallocate()*
        //
        // As a consequence, MatAllocator_LinkArray is created via a *naked* new() and will be destroyed via deallocate()
        //
```

**Comment by pthom** [comment]
And  register_instance + unregister_instance are here to keep alive the allocator instances for the alive cv::Mat

**Comment by pthom** [comment]
if you think you can find a simpler implementation, please do

**Comment by virtuald** [comment]
See #15

**Comment by pthom** [comment]
Fixed by #15, thanks a lot!
